packages/devtools_shared/lib/src/extensions/extension_manager.dart

```diff
@@ -0,0 +1,168 @@
+      //   'devtools',
+      //   packageConfig: '$rootPath/.dart_tool/package_config.json',
+      // );
+      final extensions = <Extension>[];
+      for (final extension in extensions) {
+        final config = extension.config;
+        if (config is! Map) {
+          // Fail gracefully. Invalid content in the extension's config.json.
+          continue;
+        }
+        final configAsMap = config as Map<String, Object?>;
+
+        // This should be relative to the 'extension/devtools/' directory and
+        // defaults to 'build';
+        final relativeExtensionLocation =
+            configAsMap['buildLocation'] as String? ?? 'build';
+
+        // TODO(kenz): this is hacky. Unclear if this will work for windows.
+        var rootWithoutFile = extension.rootUri.toString();
+        if (rootWithoutFile.startsWith('file:///Users')) {
+          rootWithoutFile = rootWithoutFile.replaceFirst('file:///', '/');
+        }
+
+        final location = path.join(
+          rootWithoutFile,
+          extensionLocation,
+          relativeExtensionLocation,
+        );
+
+        try {
+          final pluginConfig = DevToolsExtensionConfig.parse({
+            ...configAsMap,
+            DevToolsExtensionConfig.pathKey: location,
+          });
+          devtoolsExtensions.add(pluginConfig);
+        } on StateError catch (e) {
+          print(e.message);
+          continue;
+        }
+      }
+    }
+
+    _resetServedPluginsDir();
+    for (final extension in devtoolsExtensions) {
+      _moveToServedExtensionsDir(extension.name, extension.path);
+    }
+  }
+
+  void _resetServedPluginsDir() {
+    final buildDirectory = Directory(buildDir);
+    if (!buildDirectory.existsSync()) {
+      throw Exception('The build directory does not exist.');
+    }
+
+    // Destroy and recreate the 'devtools_extensions' directory where extension
+    // assets are served.
+    final servedExtensionsDir = Directory(servedExtensionsPath);
+    if (servedExtensionsDir.existsSync()) {
+      servedExtensionsDir.deleteSync(recursive: true);
+    }
+    servedExtensionsDir.createSync();
+  }
+
+  void _moveToServedExtensionsDir(
+    String extensionPackageName,
+    String extensionPath,
+  ) {
+    final newExtensionPath =
+        path.join(servedExtensionsPath, extensionPackageName);
+    copyPathSync(extensionPath, newExtensionPath);
+  }
+}
+
+/// Mark copied from package:extension_discovery. Remove once it is published.
+
+/// Information about an extension for target package.
+class Extension {
+  Extension._({
+    required this.package,
+    required this.rootUri,
+    required this.packageUri,
+    required this.config,
+  });
+
+  /// Name of the package providing an extension.
+  final String package;
+
+  /// Absolute path to the package root.
+  ///
+  /// This folder usually contains a `lib/` folder and a `pubspec.yaml`
+  /// (assuming dependencies are fetched using the pub package manager).
```

Comment:
We make references to fetching using `dart pub` below, but use "fetched using the pub package manager" verbiage here. We should probably pick one or the other (FWIW, I prefer just saying "dart pub").

Reply:
Ack. But this code is copied from https://github.com/dart-lang/tools/pull/129, and it will eventually be removed from devtools_shared. It is only in this package temporarily. I will prune the comments and make it a private class to make that more clear.